File: packages/python/src/agent_wallet/core/resolver.py
"""Wallet resolution helpers."""

from __future__ import annotations

import os
from collections.abc import Mapping
from pathlib import Path

from agent_wallet.core.base import (
    ENV_AGENT_WALLET_DIR,
    ENV_AGENT_WALLET_PASSWORD,
    Wallet,
    WalletProvider,
)
from agent_wallet.core.config import (
    WalletsTopology,
    load_config,
    load_runtime_secrets_password,
)
from agent_wallet.core.providers.config_provider import ConfigWalletProvider
from agent_wallet.core.providers.env_provider import EnvWalletProvider
from agent_wallet.core.utils.env import clean_env_value
from agent_wallet.local.secret_loader import load_local_secret

_DEFAULT_SECRETS_DIR = os.path.join(Path.home(), ".agent-wallet")
def resolve_wallet_provider(
    *,
    network: str | None = None,
    dir: str | None = None,
) -> WalletProvider:
    """Resolve and return the provider selected from config first, then env."""
    env = os.environ
    resolved_dir = _resolve_dir(dir, env)
    password = _resolve_password(resolved_dir, env)

    if password:
        return ConfigWalletProvider(
            resolved_dir,
            password,
            network=network,
            secret_loader=load_local_secret,
        )

    config = _load_config_safe(resolved_dir)
    if _has_available_config_wallet(config):
        return ConfigWalletProvider(
            resolved_dir,
            network=network,
            secret_loader=load_local_secret,
        )

    return EnvWalletProvider(
        network=network,
        env=env,
    )

async def resolve_wallet(
    *,
    network: str | None = None,
    dir: str | None = None,
    wallet_id: str | None = None,
) -> Wallet:
    """Resolve and return the active wallet from config first, then env."""
    provider = resolve_wallet_provider(network=network, dir=dir)
    if isinstance(provider, ConfigWalletProvider):
        return (
            await provider.get_wallet(wallet_id, network)
            if wallet_id
            else await provider.get_active_wallet(network)
        )
    if isinstance(provider, EnvWalletProvider):
        return await provider.get_active_wallet(network)
    raise ValueError(f"Unsupported provider resolved: {type(provider).__name__}")



def _resolve_dir(dir: str | None, env: Mapping[str, str]) -> str:
    if dir:
        return os.path.expanduser(dir)

    env_dir = clean_env_value(env, ENV_AGENT_WALLET_DIR)
    if env_dir:
        return os.path.expanduser(env_dir)

    return _DEFAULT_SECRETS_DIR


def _resolve_password(secrets_dir: str, env: Mapping[str, str]) -> str | None:
    file_password = load_runtime_secrets_password(secrets_dir)
    if file_password:
        return file_password
    return clean_env_value(env, ENV_AGENT_WALLET_PASSWORD)


def _load_config_safe(secrets_dir: str) -> WalletsTopology | None:
    try:
        return load_config(secrets_dir)
    except FileNotFoundError:
        return None


def _has_available_config_wallet(config: WalletsTopology | None) -> bool:
    return bool(config and config.wallets)

File: packages/typescript/src/core/resolver.ts
/**
 * Wallet resolution helpers.
 */

import { homedir } from 'node:os'
import { join } from 'node:path'

import type { Wallet } from './base.js'
import { ENV_AGENT_WALLET_DIR } from './base.js'
import {
  ConfigNotFoundError,
  loadConfig,
  type WalletsTopology,
} from './config.js'
import { ConfigWalletProvider } from './providers/config-provider.js'
import { EnvWalletProvider } from './providers/env-provider.js'
import { cleanEnvValue } from './utils/env.js'

const DEFAULT_SECRETS_DIR = join(homedir(), '.agent-wallet')

export type ResolvedWalletProvider = ConfigWalletProvider | EnvWalletProvider

export function resolveWalletProvider(options?: {
  network?: string
  dir?: string
}): ResolvedWalletProvider {
  const resolvedDir = resolveDir(options?.dir)

  const config = loadConfigSafe(resolvedDir)
  if (hasAvailableConfigWallet(config)) {
    return new ConfigWalletProvider(resolvedDir, {
      network: options?.network,
    })
  }

  return new EnvWalletProvider({
    network: options?.network,
  })
}

export async function resolveWallet(options?: {
  network?: string
  dir?: string
  walletId?: string
}): Promise<Wallet> {
  const provider = resolveWalletProvider({ network: options?.network, dir: options?.dir })

  if (provider instanceof ConfigWalletProvider) {
    return options?.walletId
      ? provider.getWallet(options.walletId, options?.network)
      : provider.getActiveWallet(options?.network)
  }

  if (provider instanceof EnvWalletProvider) {
    return provider.getActiveWallet(options?.network)
  }

  throw new Error(`Unsupported provider resolved: ${(provider as object).constructor.name}`)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function expandTilde(p: string): string {
  if (p === '~' || p.startsWith('~/') || p.startsWith('~\\')) return join(homedir(), p.slice(2))
  return p
}

function resolveDir(dir: string | undefined): string {
  if (dir) return expandTilde(dir)
  const envDir = cleanEnvValue(process.env[ENV_AGENT_WALLET_DIR])
  if (envDir) return expandTilde(envDir)
  return DEFAULT_SECRETS_DIR
}

function loadConfigSafe(secretsDir: string): WalletsTopology | null {
  try {
    return loadConfig(secretsDir)
  } catch (error) {
    if (!(error instanceof ConfigNotFoundError)) {
      throw error
    }
    return null
  }
}

function hasAvailableConfigWallet(config: WalletsTopology | null): boolean {
  return Boolean(config && Object.keys(config.wallets).length > 0)
}
